src/grpc/utils/utils.ts

```diff
@@ -482,6 +482,59 @@ function promisifyDuplexStreamCall<TRead, TWrite>(
   };
 }
 
+/**
+ * This adds a watcher for connection state change that logs the state transitions.
+ */
+function watchForChannelClose(
+  client: Client,
+  destructionCallback: () => Promise<void>,
+  logger: Logger = new Logger(watchForChannelClose.name),
+) {
```

Comment:
This function appears to be a procedure. And since it is only used in `GRPCClient`, I believe this should be embedded in `GRPCClient`. That way there's no need for this awkward `logger` parameter.

(I know I use it in the tests, but that's an exception to the rule).

Reply:
So some notes from discussion:

1. Move this back into `GRPCClient` as a method (possibly static method).
2. Use the ENUMs when checking the connectivity state `grpc.connectivityState.*`.
3. Don't just catch errors or `if (err)`, you need to check if the error is an instance of the expected exception, and if it is not either throw the error, or we can do something similar to the `handleError` that occurs under the `network/ConnectionForward.ts`.
4. Don't use `setImmediate`, just register the handler immediately. No need to make it asynchronous.
5. Some of the logger messages can be quite noisy. They should be debug messages if necessary.
6. Change the name to be `watchChannelState`.

Reply:
The current architecture of this is quite complex.

An alternative is to pass a `destroyCallback` down from NCM to NC, to GCA and GC. And this `destroyCallback` is actually called by the `destroy()` at `GC`.

The NC can augment the `destroyCallback` with additional operations prior.

Then the `watchConnectivityState` can call `destroy()` on GC directly instead of having a separate path towards destroying things.

This would have be done in the GCA since you don't have access to the GC instance until you do your async creation. This should work as long as there's no further async ops after waiting for a ready connection. Otherwise the GC/GCA would need to be a StartStop pattern.

Reply:
The advantage here is that it avoids race conditions between the `channelCloseCallback` and `destroy`. Any side-effects always occur in a deterministic order.

Reply:
I've mostly implemented this now. However there is a deadlock problem between the NCM and the NC when NCM.destroyConnection() -> NC.destroy() -> NC.destroyCallback() -> NCM.destroyConnection(). calling destroy on something that is destroyed is a NOP, but it still locks when it's in the process of destroying. if we have a cycle like this then it is a deadlock. I've fixed this between NodeConnection and GRPCClientAgent by checking if its destroying. But I'm not sure that's the best way so take a look at that.

Reply:
The refactored `destroyCallback` is working now. I fixed the deadlocks by checking if the we're already destroying before calling `destroy()` in the callbacks.